### PR TITLE
feat(thicktoken): replace tiktoken with in-repo Rust→WASM tokenizer engine

### PR DESCRIPTION
**Pull request by slvnperron**
## Summary

Swaps thicktoken's engine from `tiktoken` to our own Rust→WASM build (`thicktoken/wasm/`) while keeping the `TextTokenizer` / `getWasmTokenizer()` public API unchanged — llmz and other consumers work as-is. The BPE core is the [tokie](https://github.com/feyninc/tokie) crate (MIT © Chonkie, Inc. — attribution in `NOTICE`), wrapped with a thin wasm-bindgen layer and a typed TS API.

## Benchmarks — before vs after

Measured on the built `dist/`, same machine (darwin/arm64, Node 22), ` TOKEN`×10M input (60 MB, exactly 10M tokens):

| op | before (tiktoken) | after | speedup |
| --- | ---: | ---: | ---: |
| `count` (new default: approximate) | 5,546 ms | **163 ms** (±0.2% error) | **~34×** |
| `count` (exact) | 5,546 ms | 1,119 ms | **~5×** |
| `split` | 8,323 ms | 3,256 ms | **~2.6×** |
| `truncate` to a small budget (10M → 1,000 tokens) | ~7,800 ms | **146 ms** | **~53×** |
| cold-start init (median) | ~54 ms | ~62 ms | ≈ par |
| bundle (single-file ESM) | 2,483 KB | **1,156 KB** | **−53%** |

Two of these are algorithmic, not just engine speed:
- **`count` is approximate by default on large inputs** (`{ approximate: false }` opts out): it tokenizes ~48 KB of stratified sample windows regardless of input size and extrapolates — cost is ~constant in input length, so the gap grows with input size (2–3 orders of magnitude on multi-MB texts).
- **`truncate` is windowed and exact**: it only tokenizes the prefix/suffix window it needs, so trimming a huge text to a small budget no longer tokenizes the whole thing. (It's also boundary-exact — the old chunked implementation dropped a token at chunk boundaries.)

## How the wasm stays small & fast to init

- Vocab is shipped **merges-only** (469 KB gz): `model.vocab` is fully derivable for a topological BPE, so it's rebuilt at init (base-256 byte-level alphabet + merge concatenations + specials).
- The tokenizer is built **directly from merge id-pairs** (no `tokenizer.json` round-trip) using tokie's `Simple` encoder (no Aho-Corasick automaton) → ~44 ms engine init, on par with tiktoken.
- Verified **token-id-identical to tiktoken** on prose, source code, unicode, contractions (independent ground truth: `gpt-tokenizer`, kept as devDependency for tests). Known alpha edge case, documented in tests: glued mid-word contractions (e.g. `na'transl`) tokenize to different ids (same text round-trip); never observed on real prose/code.

## Vocab variants (edge targets)

One code-only wasm engine (~127 KB gz); each entry bundles its own vocab asset (truncated cl100k merge lists — valid BPE by prefix-closure). Same `TextTokenizer` API everywhere:

| import | vocab | total gz (ESM) | init | counts vs cl100k |
| --- | --- | ---: | ---: | --- |
| `@bpinternal/thicktoken` | cl100k | ~647 KB | ~60 ms | exact |
| `@bpinternal/thicktoken/lite` | cl50k | ~392 KB | ~25 ms | overcounts ~+3-7% |
| `@bpinternal/thicktoken/micro` | cl25k | ~275 KB | ~13 ms | overcounts ~+8-20% |

Truncated variants only ever **overcount** (safe for budget enforcement — never overflows a window); not for billing math. Inflation is content-dependent (worst on emoji/unicode-heavy text). Assets regenerate via `wasm/scripts/gen-assets.mjs`; all variants verified in Chromium (`browser-test.html`).

## API changes (⚠️ minor break → v3.0.0)

- `count(text, max?: number)` → `count(text, options?: { approximate?: boolean })`. The old `max` ceiling is gone (obsolete — counting is fast now). llmz only calls `count(text)`, unaffected.
- New exports: `WasmTokenizer` (direct engine access: `encode`/`decode`/`split`/`slice`/`truncate(text, n, 'head'|'tail'|'middle')`), `TruncateMode`, `CountOptions`.
- `tiktoken` dependency removed.

## Rebuilding the wasm

`wasm/pkg/` is committed, so the package builds and tests **without a Rust toolchain**. To rebuild from Rust source: see `wasm/README.md` (wasm-pack, `--target web`).

## Test plan

- [x] 63 tests pass: full legacy `TextTokenizer` suite + 39 primitive tests (parity vs independent cl100k reference, exact truncate head/tail/middle, slice incl. negative indices, approximate-count error bounds, documented edge case)
- [x] `tsc --noEmit` clean
- [x] CJS + ESM dist smoke-tested (init, count, split, truncate, splitAndSlice)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: thicktoken/wasm/index.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+   * means "to the end". Exact.
+   */
+  slice(text: string, start = 0, end?: number): string {
+    // Resolve negative / omitted indices against the exact token count.
+    const total = this.raw.count(text, false)
+    const s = start < 0 ? Math.max(0, total + start) : Math.min(start, total)
+    const e = end === undefined ? total : end < 0 ? Math.max(0, total + end) : Math.min(end, total)
+    if (e <= s) return ''
+    return this.raw.slice(text, s, e)
+  }
```
[the remark this answers is not in the document]
Comment (slvnperron) [reply thicktoken/wasm/index.ts:99]: Fixed in e852290 — went one step further than suggested: negative/omitted index resolution now happens inside the wasm against the single encode pass, so `slice` is one full tokenization for every index form (the JS wrapper no longer calls `count` at all).

File: thicktoken/src/tokenizer.ts [hunk 2]
```diff
@@ -1,44 +1,34 @@
...
 
-const CHUNK_SIZE = 100_000
+export type CountOptions = {
+  /**
+   * When true (default), very large inputs are counted by statistical sampling —
+   * orders of magnitude faster and within a few %. Small inputs
+   * are always counted exactly. Pass false to force an exact count.
+   */
+  approximate?: boolean
+}
```
[the remark this answers is not in the document]
Comment (slvnperron) [reply thicktoken/src/tokenizer.ts]: Fixed in e852290 — `CountOptions` is now defined once in `wasm/index.ts` and re-exported through `src/core.ts` and all three entries.

**Review by franklevasseur — APPROVED** [review]
